### ISSUE TITLE
Inserting a poll into an article using shortcode [sexypolling id="id"]

### ISSUE THREAD

**Issue by JindraEkart** [Jefferson49/joomla4_plugin_sexy_polling#86]
I installed Sexy Polling Reloaded, v5.0.4.0 (for Joomla 4.x, 3.10) into Joomla version 3.10.12.

I set up a test poll in the module and in the article using a shortcode.

Everything works correctly in the module:  
URL: [https://testnove.silvarium.net/zpravy-z-oboru-lesnictvi-a-drevarstvi](url)
The only problem is that it doesn't load the width: 100% parameter from the poll settings, but instead sets a fixed width of the poll block, and that's not correct for a responsive website. This can be solved by overriding it in CSS.

In the article, the CSS for the poll template does not load, and when the results are displayed, the graph bars are not generated, color.js error.
URL: [https://testnove.silvarium.net/lesnictvi/budoucnost-je-ve-vetsim-propojeni-lesnictvi-a-zpracovani-drivi](url)

_Cannot read properties of undefined (reading '0')
     at jQuery.fx.step.<computed> (color.js:24:64)
     at fx.update (sexylib.js:4:22881)
     at fx.step (sexylib.js:4:24820)
     at h (sexylib.js:4:23149)
     at fx.custom (sexylib.js:4:23527)
     at HTMLDivElement.g (sexylib.js:4:21163)
     at Function.dequeue (sexylib.js:2:24179)
     at HTMLDivElement.<anonymous> (sexylib.js:2:24508)
     at Function.each (sexylib.js:2:11776)
     at init.each (sexylib.js:2:8304)_

Best regards
Jindrich Ekart

**Comment by Jefferson49** [comment]
Hello Jindrich,

thank you for reporting your issues.

> Everything works correctly in the module:
URL: [https://testnove.silvarium.net/zpravy-z-oboru-lesnictvi-a-drevarstvi](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/issues/url)
The only problem is that it doesn't load the width: 100% parameter from the poll settings, but instead sets a fixed width of the poll block, and that's not correct for a responsive website. This can be solved by overriding it in CSS.

I am not sure about the width topic. For me, it looks like in the screenshot below, which seems to be o.k.. Does it look different in your browser? Can you mark the wrong width in a screenshot?


![silvarium1](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/81484983/21a429bb-fb78-4998-8fc6-83b9d8a76c4a)

> doesn't load the width: 100% parameter from the poll settings

I am not sure, which parameter do you refer to. A parameter in the template settings? Can you describe in more detail?

> This can be solved by overriding it in CSS

Can you paste the CSS code, which seems to be wrong? What did you change to override it?

**Comment by Jefferson49** [comment]
> In the article, the CSS for the poll template does not load, and when the results are displayed, the graph bars are not generated, color.js error.
URL: [https://testnove.silvarium.net/lesnictvi/budoucnost-je-ve-vetsim-propojeni-lesnictvi-a-zpracovani-drivi](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/issues/url)

I can reproduce this. For me, it looks like this:

![silvarium_bug2](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/81484983/e1023d9a-8155-4d68-88bc-547a751e24e0)

Maybe, this issue looks a little bit simular to earlier issues #62, #63. They occurred, if a group of votes (i.e. "Polls" instead of "Single Poll") was shown in one site module. However, should be fixed in version 5.0.4.0. Are you sure you use 5.0.4.0?

To analyze the background, it would be very helpful if you could send me the Sexypolling tables from your database. These tables contain all the configuration and poll data from your installation. You can select the 5 tables in phpMyAdmin and export them to a SQL file. The 5 tables are shown in the following screenshot:

![293150893-bd5ae9a7-1082-4d7b-8bdc-a2d69913ef85](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/81484983/b5f94c87-4015-4f73-9f64-7c8dbcc58989)

You can ZIP the SQL file and attach it to this issue. If you do not want to share your data on Github, you can also send the SQL file by email to: sexypolling.reloaded(at)gmail.com

**Comment by Jefferson49** [comment]
From your screenshot, I just saw that some of the Czech translation is missing:

![translation](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/81484983/5bde8a69-47fb-4348-8ebf-9e2920fd0cd4)

You can create a Czech language file and add it to the installation zip file:
com_sexypolling\language\cs-CZ\cs-CZ.com_sexypolling.ini

You can copy an existing language file from \en-GB\en-GB.com_sexypolling.ini and rename it. Afterwards change the translation. 

You also need to add the language file to the component XML file: com_sexypolling\sexypolling.xml
```
<languages folder="site">
        <language tag="cs-CZ">language/cs-CZ/cs-CZ.com_sexypolling.ini</language>
</languages>

```

If you would like to contribute to the project, you can send me the translation file and I will add it to the module code.

**Comment by Jefferson49** [comment]
> color.js error

> Cannot read properties of undefined (reading '0')
> at jQuery.fx.step. (color.js:24:64)

From the information in this error message, the error seem to occur in the following line of code:
https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/blob/joomla_5.x/com_sexypolling/site/assets/js/color.js#L24

Seemingly, some RGB values are calculated.

Could this be some color value from your template? Which template did you use? Did you change some color values?

**Comment by JindraEkart** [comment]
Hello

my version of component is 5.0.4.0:
![version](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/161603675/3dea438f-f89c-4e03-b852-9bee7a34912b)

there is part of settings for test poll:  Poll width: 100%
![poll-style-options](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/161603675/ad855616-5bce-4578-9c50-9a87c775d188)

but the module will be generated with a fixed width of 353px:
![module poll code css](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/161603675/cea75ed4-5379-438c-97ff-366b8cf7a67e)

I solved this by overriding my own CSS:
.polling_container_wrapper {
    width: 100%!important;
}

The problem with the embedded poll using the shortcode is because the CSS file for the poll settings is not generated, it is this part of the code in the HEAD section:

<link rel="stylesheet" href="[/index.php?option=com_ajax&module=sexypolling&method=generateCSS&format=raw?id_poll=2&module_id=10000](https://testnove.silvarium.net/index.php?option=com_ajax&module=sexypolling&method=generateCSS&format=raw?id_poll=2&module_id=10000)" type="text/css" />

![CSS for poll in article](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/161603675/07297458-8de3-4269-9ee9-a42231de4d06)

![poll in article](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/161603675/2857c7bd-91cf-49b9-bd97-6693b3efe73a)


 I am attaching the table export for the component:
[tables-sexypollings.zip](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/files/14445375/tables-sexypollings.zip)

Best regards
Jindřich Ekart

**Comment by Jefferson49** [comment]
Thanks a lot for the excellent explanation and screenshots! That is a huge help.

 I can reproduce the "Poll width: 100%" issue on my own Joomla system with my own poll example. I also get a fixed value. At the moment, I do not understand why it does not use the setting from the component administration, because the code looks alright. 

It will probably need some time to check in a debugging session.

**Comment by Jefferson49** [comment]
> The problem with the embedded poll using the shortcode is because the CSS file for the poll settings is not generated, it is this part of the code in the HEAD section:

The module uses an AJAX communication, which can also be seen from your screenshot (href="/index.php?option=com_ajax..."). With AJAX, a function in the Sexy Polling PHP code is called, which generates and returns the CSS.

For the AJAX communication, the Joomla Ajax Interface component (com_ajax) is used, which should be part of every Joomla installation.

Can you check if this component is available and activated in your Joomla installation? Could there be any other reason, why a http-call to this component would not work?

![com_ajax](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/81484983/01e2d100-356a-4ad6-aef9-73730e375191)

**Comment by JindraEkart** [comment]
It looks like the Ajax component is installed in here:
![ajax component](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/161603675/552947d5-e672-49ad-a00b-935e594c858d)

**Comment by JindraEkart** [comment]
I get this error when I try to activate the CSS generation link in HEAD:
[https://testnove.silvarium.net/index.php?option=com_ajax&module=sexypolling&method=generateCSS&format=raw?id_poll=2&module_id=10000](url)

![error-ajax](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/161603675/eadd1110-a9b3-47c7-a54c-eb2fc92f2962)

This means:: LogicException: Method generateCSSAjax does not exist.

According to this guide: [https://docs.joomla.org/Using_Joomla_Ajax_Interface](url)
There should be a method (function) called generateCSSAjax in the helper.php file in the mod_sexypolling module, but it is not there.

**Comment by Jefferson49** [comment]
> There should be a method (function) called generateCSSAjax in the helper.php file in the mod_sexypolling module, but it is not there.

Yes, this is a bug. If the Sexypolling **plugin** is used, the generateCSS Ajax method is called "**onAfterRender**" and the related method is not available. I attached a new version of the extension, which is based on v5.0.4.0 and includes a bugfix.

[com_sexypolling_5.0.4.0_4e628417.zip](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/files/14456753/com_sexypolling_5.0.4.0_4e628417.zip)

However, I cannot test the bugfix, because I cannot reproduce the behavior in my own installation.

Can you paste the exact way how you include the module in your site? For example, something like include "{loadmoduleid xyz}" in an article?

Do you use any specific things around the SexyPolling **plugin** and **onAfterRender**?

Could you install the bugfix and test if it fixes your issue?

**Comment by Jefferson49** [comment]
> However, I cannot test the bugfix, because I cannot reproduce the behavior in my own installation.

Now, I could reproduce the issue by inserting [sexypolling id="1"] into an article. From the Github issue title, I guess that's the same way how you use the module. Seemingly, this triggers the plugin and causes the specific behavior.

Seemingly, the attached bugfix is not sufficient. I will need to further investigate this issue.

As a workaround, you can load the module in the article in a different way: {loadmoduleid xyz}

**Comment by JindraEkart** [comment]
Hello

yes, as written in the title of the issue, the article contains a poll using the shortcode [sexypolling id="2"].
I thought this shortcode was meant to be inserted into an article using a plugin.

![shortcode](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/assets/161603675/05ef7a0a-f665-43f1-912d-94f18a666056)

I know that it is possible to insert a module into the article, but I am looking for polls that can be inserted into the article using a plugin (shortcode).
We are looking for a solution that is suitable for editors who are not very experienced in coding.
If the module were to be inserted into the article, someone must first create a module for each poll, and that is not very practical with a large number of polls.

We run a news site where multiple editors post articles and we are looking for a replacement for this component (Advanced Poll by Extore): [https://extstore.com/joomla-extensions/components/advanced-pol](url)

This component has not been updated for a long time and we have a problem with jQuery conflict, the polls do not work reliably for us.

I also have an idea for improving the Sexy polling component:
1. The option to copy the poll (or copy the poll settings) is missing here, it will speed up the setting of the same or similar poll.
2. In the global settings, there could be more default settings for polls, so that every new poll would not have to be set up again.
The new poll would take the default settings from the global settings.
3. If the shortcode poll embedding issue is resolved, a button plugin for the editor to embed the poll into the article would help.
4. Option to turn off animations in polls (global property, poll property)
5. Possibility to turn off the reordering of results according to the number of poll votes, not everyone wants it (global property, poll property)

Best regards
Jindrich Ekart

**Comment by Jefferson49** [comment]
Now, I managed to fixed the issue with shortcode [sexypolling id="id"].

If you want to test, you can use the attached version.

[com_sexypolling_v5.0.4.0_5890b4e3.zip](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/files/14469015/com_sexypolling_v5.0.4.0_5890b4e3.zip)

**Comment by Jefferson49** [comment]
> I can reproduce the "Poll width: 100%" issue on my own Joomla system with my own poll example. I also get a fixed value. At the moment, I do not understand why it does not use the setting from the component administration, because the code looks alright.

I created a separate issue for this => #87

**Comment by Jefferson49** [comment]
> The option to copy the poll (or copy the poll settings) is missing here, it will speed up the setting of the same or similar poll.

I created a new issue for an enhancement => 88

**Comment by Jefferson49** [comment]
> In the global settings, there could be more default settings for polls, so that every new poll would not have to be set up again.
The new poll would take the default settings from the global settings.

That would be more difficult, since the poll settings follow a certain standard Joomla structure. The shown settings are generated by the Joomla framework from this XML file:
https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/blob/joomla_5.x/com_sexypolling/admin/models/forms/sexypoll.xml

The default settings for each new poll are contained as strings in this XML file. It would need a new concept to change that.

If copying of polls would be possible, a "default" poll could be created and all inherited polls could be created as a copy of this default poll? Do you think this would be sufficient?

**Comment by Jefferson49** [comment]
> If the shortcode poll embedding issue is resolved, a button plugin for the editor to embed the poll into the article would help.

I have no experience in this area. Is there some API available to add buttons to the WYSIWIG editor for articles?

**Comment by Jefferson49** [comment]
> Option to turn off animations in polls (global property, poll property)

> Possibility to turn off the reordering of results according to the number of poll votes, not everyone wants it (global property, poll property)

Those issues are mostly performed in the JavaScript part of the code. In a brief check, I did not see a configuration parameter which is handed over from PHP to JavaScript.

Since starting with the maintenance of SexyPolling, I followed the strategy to keep the JavaScript as it is.

**Comment by Jefferson49** [comment]
Reopened to continue discussion

**Comment by Jefferson49** [comment]
The topic about the width is a little bit weired. The width is correctly handled within PHP and the html is generated with width=100%. However, after loading the page, a [function in JavaScript](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/blob/joomla_5.x/com_sexypolling/site/assets/js/sexypolling_unpacked.js#L73) overrides this setting and inserts the current width in pixels:

```
                $(".polling_container_wrapper").each(function() {
                    var w = $(this).width();
                    $(this).css("width",w);
                    var curr_h = $(this).find('.answer_wrapper').height();
                    $(this).find('.answer_wrapper').attr("h",curr_h);
                });
```

If I delete the line with $(this).css("width",w), the width keeps at 100%

At the moment, I do not know what is best. Deleting seems to help, but creates the risk that this line of code has some intention and will lead to unintended behavior if deleted. However, so far there seems to be no effect.

**Comment by JindraEkart** [comment]
Hello

I installed the patched version on the test site: com_sexypolling_v5.0.4.0_5890b4e3.zip.
Now the CSS code for the poll is generated fine, so the poll is working: [https://testnove.silvarium.net/lesnictvi/budoucnost-je-ve-vetsim-propojeni-lesnictvi-a-zpracovani-drivi](url)

It's also great news that a new feature will be added - poll copy.

Thank you for the excellent work.

Best regards
Jindřich Ekart

**Comment by Jefferson49** [comment]
Regarding copying of polls, I managed to include "Save as Copy" in the poll view, which was quite simple. I also tried to add "Copy" in the polls view, which is not working yet and is more complicated than expected. See screenshot in #88.

The attachement contains a version with bugfix for width and "Save as Copy" in the poll view.

[com_sexypolling_v5.0.4_dfd51377.zip](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/files/14490930/com_sexypolling_v5.0.4_dfd51377.zip)

**Comment by JindraEkart** [comment]
Hello

It is not possible download this file (The attachement contains a version with bugfix for width and "Save as Copy" in the poll view):
com_sexypolling_v5.0.4_dfd51377.zip <https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/files/14490930/com_sexypolling_v5.0.4_dfd51377.zip> 

Best regards
Jindrich Ekart

**Comment by Jefferson49** [comment]
> It is not possible download this file 

I do not know what happened here. Maybe, a temporary problem, because it works for me.

In the mean time, I was able to add a copy feature to the poll view and to the polls view as well. The attachement contains the copy features and the width bugfix. I hope it works this time.

[com_sexypolling_v5.0.4_dca23b82.zip](https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/files/14501464/com_sexypolling_v5.0.4_dca23b82.zip)

**Comment by JindraEkart** [comment]
Hello

 

Yes, the file has already been downloaded.

 

Copying the poll works fine.

Thank you, very good work :)

 

I will also process the Czech translation for this component and send it to you.

 

Best regards
Jindrich Ekart

**Comment by Jefferson49** [comment]
Implemented the issue in title (#86) with https://github.com/Jefferson49/Joomla_plugin_sexypolling_reloaded/commit/5890b4e3d5b3a226b9e19ab7083b7022aa47c789. 

Solved further issues and proposed new features within separate issues: #87, #88, #89.

**Comment by Jefferson49** [comment]
I published a new release 5.0.5.0/5.5.0, which includes the new features as discussed.
